### PR TITLE
Add Falcon single-gpu

### PR DESCRIPTION
**Pull request by wtmlon**
### Description

'''
import paddle
from paddlenlp.transformers import RWForCausalLM, RWTokenizer

tokenizer = RWTokenizer.from_pretrained("tiiuae/falcon-7b")
paddle_model = RWForCausalLM.from_pretrained("tiiuae/falcon-7b", dtype="float16")
paddle_model.eval()
input_features = tokenizer("Hello! Please introduce yourself.", return_tensors="pd")
with paddle.no_grad():
    paddle_outputs = paddle_model.generate(**input_features, max_length=100)
    print(tokenizer.batch_decode(paddle_outputs[0].tolist()))
#['\nMy name is Karla and I am a 30-year-old mother of two. I am a stay-at-home mom and I love it! I am also a wife to a wonderful man who is a pastor. I am a Christian and I love the Lord with all my heart. I am a lover of all things creative and I love to write. I am a lover of all things Disney and I love to travel. I am a lover of all things family and I']

tokenizer = RWTokenizer.from_pretrained("tiiuae/falcon-7b-instruct")
paddle_model = RWForCausalLM.from_pretrained("tiiuae/falcon-7b-instruct", dtype="float16")
paddle_model.eval()
input_features = tokenizer("Hello! Please introduce yourself.", return_tensors="pd")
with paddle.no_grad():
    paddle_outputs = paddle_model.generate(**input_features, max_length=100)
    print(tokenizer.batch_decode(paddle_outputs[0].tolist()))

#["\nHello! I'm a 20 year old girl from the UK. I'm a student studying psychology and I'm currently doing my dissertation on the topic of mental health. I'm interested in learning about different cultures and meeting new people. I'm a big animal lover and I love to go on walks in the countryside. I'm also a big fan of music and I love to go to gigs and festivals. I'm looking forward to learning about different cultures and"]
'''

### CODE REVIEW

File: examples/language_model/rw/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Falcon-7B
```
Comment (sijunhe) [inline examples/language_model/rw/README.md]: 写Falcon就好，他也不止7b一个模型和大小

File: paddlenlp/transformers/rw/__init__.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/__init__.py]: 所有的copyright加上原作者

File: paddlenlp/transformers/rw/modeling.py [hunk 3]
```diff
@@ -0,0 +1,1127 @@
...
+
+# NOTE(Hesslow): Unfortunately we did not fuse matmul and bias during training, this means that there's one additional quantization to bfloat16 between the operations.
+# In order not to degrade the quality of our HF-port, we keep these characteristics in the final model.
+class Linear(nn.Linear):
+    def forward(self, input: Tensor) -> Tensor:
+        ret = input @ self.weight
+        if self.bias is None:
+            return ret
+        else:
+            return ret + self.bias
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 这个linear是不是非加不可？用普通的nn.linear能否对其？

File: paddlenlp/transformers/rw/modeling.py [hunk 4]
```diff
@@ -0,0 +1,1127 @@
...
+            attn_output = query_layer_ @ key_layer_.transpose([0, 1, 3, 2])
+            attention_scores = attn_output.reshape([batch_size, self.num_heads, q_length, kv_length])
+
+
+            # cast attention scores to fp32, compute scaled softmax and cast back to initial dtype - [batch_size, num_heads, q_length, kv_length]
+            input_dtype = attention_scores.dtype
+            # `float16` has a minimum value of -65504.0, whereas `bfloat16` and `float32` have a minimum value of `-3.4e+38`
+            if input_dtype == paddle.float16 or input_dtype == paddle.bfloat16:
+                attention_scores = paddle.cast(attention_scores, paddle.float32)
+            attention_scores = paddle.where(attention_mask > 0, paddle.full_like(attention_scores, -1e38), attention_scores)
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]:
```suggestion
            attention_scores = paddle.where(attention_mask > 0, paddle.full_like(attention_scores, paddle.finfo(attention_scores.dtype).min), attention_scores)
```

File: paddlenlp/transformers/rw/modeling.py [hunk 5]
```diff
@@ -0,0 +1,1127 @@
...
+            query_layer_ = query_layer.reshape([batch_size, self.num_heads, q_length, self.head_dim])
+            key_layer_ = key_layer.reshape([batch_size, self.num_kv, kv_length, self.head_dim])
+            value_layer_ = value_layer.reshape([batch_size, self.num_kv, kv_length, self.head_dim])
+
+            alibi = alibi.reshape([batch_size, self.num_heads, 1, -1])
+
+            attention_scores = query_layer_ @ key_layer_.transpose([0, 1, 3, 2])
+
+            attention_mask_float = paddle.zeros_like(attention_mask, dtype=attention_scores.dtype)
+            attention_mask_float = paddle.where(attention_mask, paddle.full_like(attention_scores, -1e9), attention_mask_float)
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]:
```suggestion
            attention_mask_float = paddle.where(attention_mask, paddle.full_like(attention_scores, paddle.finfo(attention_scores.dtype).min), attention_mask_float)
```

File: paddlenlp/transformers/rw/modeling.py [hunk 6]
```diff
@@ -0,0 +1,1127 @@
...
+        if use_cache:
+            outputs = (output,) + outputs
+        else:
+            outputs = (output,) + outputs[1:]
+
+        return outputs  # hidden_states, present, attentions
+
+
+class RWPreTrainedModel(PretrainedModel):
+    _keys_to_ignore_on_load_missing = [r"h.*.self_attention.scale_mask_softmax.causal_mask", r"lm_head.weight"]
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 这个有用到吗

File: paddlenlp/transformers/rw/modeling.py [hunk 7]
```diff
@@ -0,0 +1,1127 @@
...
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = RWConfig
+    base_model_prefix = "transformer"
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["DecoderLayer"]
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: paddle这边没这几个

File: paddlenlp/transformers/rw/modeling.py [hunk 8]
```diff
@@ -0,0 +1,1127 @@
...
+                        ["classifier.bias", None],
+                    ]
+                )
+            if "RWForQuestionAnswering" in config.architectures:
+                mappings.extend(
+                    [
+                        ["qa_outputs.weight", None, "transpose"],
+                        ["qa_outputs.bias", None],
+                    ]
+                )
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 这几个可以不加

File: paddlenlp/transformers/rw/modeling.py [hunk 9]
```diff
@@ -0,0 +1,1127 @@
...
+
+from ..model_outputs import (
+    BaseModelOutputWithPastAndCrossAttentions,
+    CausalLMOutputWithCrossAttentions,
+    QuestionAnsweringModelOutput,
+    SequenceClassifierOutputWithPast,
+    TokenClassifierOutput,
+)
+
+from paddle.nn import BCEWithLogitsLoss, CrossEntropyLoss, LayerNorm, MSELoss
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 这些统一直接用`nn.LayerNorm`, 不需要再import一次了

File: paddlenlp/transformers/rw/modeling.py [hunk 10]
```diff
@@ -0,0 +1,1127 @@
...
+                    use_cache = False
+
+                def create_custom_forward(module):
+                    def custom_forward(*inputs):
+                        # None for past_key_value
+                        return module(*inputs, use_cache=use_cache, output_attentions=output_attentions)
+
+                    return custom_forward
+
+            else:
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 这块gradient_checkpointing的逻辑如果是torch那边直接来的，就删掉, 我们这边写法不一样

File: paddlenlp/transformers/rw/modeling.py [hunk 11]
```diff
@@ -0,0 +1,1127 @@
...
+        """
+        if deprecated_arguments.pop("position_ids", False) is not False:
+            # `position_ids` could have been `Tensor` or `None` so defaulting pop to `False` allows to detect if users were passing explicitly `None`
+            warnings.warn(
+                "`position_ids` have no functionality in BLOOM and will be removed in v5.0.0. You can safely ignore"
+                " passing `position_ids`.",
+                FutureWarning,
+            )
+        if len(deprecated_arguments) > 0:
+            raise ValueError(f"Got unexpected arguments: {deprecated_arguments}")
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 这些就没必要了吧

File: paddlenlp/transformers/rw/modeling.py [hunk 12]
```diff
@@ -0,0 +1,1127 @@
...
+        hidden_states = transformer_outputs[0]
+
+        lm_logits = self.lm_head(hidden_states)
+
+        loss = None
+        if labels is not None:
+            # Shift so that tokens < n predict n
+            shift_logits = lm_logits[..., :-1, :]
+            shift_labels = labels[..., 1:]
+            batch_size, seq_length, vocab_size = shift_logits.shape
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 参考 https://github.com/PaddlePaddle/PaddleNLP/pull/6328 修改

File: paddlenlp/transformers/rw/modeling.py [hunk 13]
```diff
@@ -0,0 +1,1127 @@
...
+    @staticmethod
+    def _reorder_cache(past: Tuple[Tuple[Tensor]], beam_idx: Tensor) -> Tuple[Tuple[Tensor]]:
+        """
+        This function is used to re-order the `past_key_values` cache if [`~PreTrainedModel.beam_search`] or
+        [`~PreTrainedModel.beam_sample`] is called. This is required to match `past_key_values` with the correct
+        beam_idx at every generation step.
+        """
+        return tuple(tuple(past_state.index_select(0, beam_idx) for past_state in layer_past) for layer_past in past)
+
+class RWForTokenClassification(RWPreTrainedModel):
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 删

File: paddlenlp/transformers/rw/modeling.py [hunk 14]
```diff
@@ -0,0 +1,1127 @@
...
+
+        return TokenClassifierOutput(
+            loss=loss,
+            logits=logits,
+            hidden_states=transformer_outputs.hidden_states,
+            attentions=transformer_outputs.attentions,
+        )
+
+
+class RWForQuestionAnswering(RWPreTrainedModel):
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 删

File: tests/transformers/rw/test_modeling.py [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+        )
+        result = model(input_ids, use_cache=True, return_dict=self.parent.return_dict)
+        self.parent.assertEqual(result[0].shape, [self.batch_size, self.seq_length, self.hidden_size])
+        #self.parent.assertEqual(result[1].shape, [self.batch_size, self.hidden_size])
+        result = model(input_ids, use_cache=True, output_attentions=True, return_dict=self.parent.return_dict)
+        self.parent.assertEqual(result[0].shape, [self.batch_size, self.seq_length, self.hidden_size])
+        #self.parent.assertEqual(result[1].shape, [self.batch_size, self.hidden_size])
+        result = model(input_ids, use_cache=True, output_attentions=True, output_hidden_states=True, return_dict=self.parent.return_dict)
+        self.parent.assertEqual(result[0].shape, [self.batch_size, self.seq_length, self.hidden_size])
+        #self.parent.assertEqual(result[1].shape, [self.batch_size, self.hidden_size])
```
Comment (sijunhe) [inline tests/transformers/rw/test_modeling.py]: 不必要的comments删

File: tests/transformers/rw/test_modeling.py [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+        model.eval()
+        result = model(
+            input_ids, attention_mask=input_mask, labels=input_ids, return_dict=self.parent.return_dict,
+        )
+        print(self.parent.return_dict, "return")
+        print(result[1], "lm logit")
+        print(result[0], "loss")
+        #if self.parent.return_dict:
+        self.parent.assertEqual(result[0].shape, [])
+        self.parent.assertEqual(result[1].shape, [self.batch_size, self.seq_length, self.vocab_size])
```
Comment (sijunhe) [inline tests/transformers/rw/test_modeling.py]: 清理

File: tests/transformers/rw/test_modeling.py [hunk 17]
```diff
@@ -0,0 +1,478 @@
...
+
+    def create_and_check_for_question_answering(
```
Comment (sijunhe) [inline tests/transformers/rw/test_modeling.py]: qa模型和单测可以删

File: tests/transformers/rw/test_modeling.py [hunk 18]
```diff
@@ -0,0 +1,478 @@
...
+        )
+        if sequence_labels is not None:
+            start_logits, end_logits = result[1], result[2]
+        else:
+            start_logits, end_logits = result[0], result[1]
+
+        self.parent.assertEqual(start_logits.shape, [self.batch_size, self.seq_length])
+        self.parent.assertEqual(end_logits.shape, [self.batch_size, self.seq_length])
+
+    def create_and_check_for_token_classification(
```
Comment (sijunhe) [inline tests/transformers/rw/test_modeling.py]: 删

File: tests/transformers/rw/test_modeling.py [hunk 19]
```diff
@@ -0,0 +1,478 @@
...
+    base_model_class = RWModel
+    return_dict: bool = False
+    use_labels: bool = False
+    use_test_inputs_embeds: bool = True
+
+    all_model_classes = (
+        RWModel,
+        RWForCausalLM,
+        RWForQuestionAnswering,
+        RWForTokenClassification,
```
Comment (sijunhe) [inline tests/transformers/rw/test_modeling.py]: 删

File: tests/transformers/rw/test_modeling.py [hunk 20]
```diff
@@ -0,0 +1,478 @@
...
+
+    def test_for_question_answering(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_for_question_answering(*config_and_inputs)
+
+    def test_for_token_classification(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_for_token_classification(*config_and_inputs)
+
+# Please execute this unittest on PaddleNLP/
```
Comment (sijunhe) [inline tests/transformers/rw/test_modeling.py]: 这个本地保留一份就行，PR上不用提交上来了

File: paddlenlp/transformers/rw/modeling.py (hunk 6 above)
Comment (wtmlon) [reply paddlenlp/transformers/rw/modeling.py]: h.*.self_attention.scale_mask_softmax.causal_mask在模型定义中不存在这样的结构，但是lm_head.weight是有的

File: paddlenlp/transformers/rw/modeling.py (hunk 3 above)
Comment (wtmlon) [reply paddlenlp/transformers/rw/modeling.py]: 可以去掉，已经做了 generate 测试，结果可以对齐

File: paddlenlp/transformers/rw/modeling.py [hunk 21]
```diff
@@ -0,0 +1,910 @@
...
+
+import paddle
+from paddle import Tensor, nn
+
+from ..model_outputs import (
+    BaseModelOutputWithPastAndCrossAttentions,
+    CausalLMOutputWithCrossAttentions,
+    QuestionAnsweringModelOutput,
+    SequenceClassifierOutputWithPast,
+    TokenClassifierOutput,
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 删

File: paddlenlp/transformers/rw/modeling.py [hunk 22]
```diff
@@ -0,0 +1,910 @@
...
+    def set_input_embeddings(self, new_embeddings: Tensor):
+        self.word_embeddings = new_embeddings
+
+    def _convert_head_mask_to_5d(self, head_mask, num_hidden_layers):
+        """-> [num_hidden_layers x batch x num_heads x seq_length x seq_length]"""
+        if head_mask.dim() == 1:
+            head_mask = head_mask.unsqueeze(0).unsqueeze(0).unsqueeze(-1).unsqueeze(-1)
+            head_mask = head_mask.expand(shape=(num_hidden_layers, -1, -1, -1, -1))
+        elif head_mask.dim() == 2:
+            head_mask = head_mask.unsqueeze(1).unsqueeze(-1).unsqueeze(-1)  # We can specify head_mask for each layer
```
Comment (sijunhe) [inline paddlenlp/transformers/rw/modeling.py]: 不用unsqueeze那么多次，直接一次到位就好了 https://www.paddlepaddle.org.cn/documentation/docs/api/paddle/unsqueeze_cn.html

**Review by sijunhe — COMMENTED** [review]
1. 剩下2个小地方修一下
2. 加入7b-instruct的ckpt

**Review by sijunhe — APPROVED** [review]
lgtm

**Pull request merged**